src/clr-angular/data/tree-view/models/tree-node.model.ts

```diff
@@ -30,22 +33,47 @@ export abstract class TreeNodeModel<T> {
   destroy() {
     // Just to be safe
     this.selected.complete();
+    this.disabled.complete();
   }
 
   // Propagate by default when eager, don't propagate in the lazy-loaded tree.
   setSelected(state: ClrSelectedState, propagateUp: boolean, propagateDown: boolean) {
-    if (state === this.selected.value) {
+    if (state === this.selected.value || this.disabled.value) {
       return;
     }
-    this.selected.next(state);
-    if (propagateDown && state !== ClrSelectedState.INDETERMINATE && this.children) {
-      this.children.forEach(child => child.setSelected(state, false, true));
+    if (propagateDown && state !== ClrSelectedState.INDETERMINATE && this.children.length > 0) {
+      this.children.filter(child => !child.disabled.value).forEach(child => child.setSelected(state, false, true));
+      // If there are any children, then we have to always calculate the selection state from the children after they are set.
+      // This is because, the state sent by parent is not always the state set on children because of the possibility of any
+      // disabled descended nodes persisting their selection states.
+      this._updateSelectionFromChildren();
+    } else {
+      this.selected.next(state);
     }
     if (propagateUp && this.parent) {
       this.parent._updateSelectionFromChildren();
     }
   }
 
+  setDisabled(value: boolean, propagateUp: boolean, propagateDown: boolean) {
+    if (this.disabled.value === value) {
+      return;
+    }
+    // If selection of the parent is disabled, then this node also has to be disabled
+    const disabledValue = this.parent && this.parent.disabled.value ? true : value;
+    this.disabled.next(disabledValue);
+    // Don't change the disableSelection state of the children for which the clrDisableSelection is set to true
+    if (propagateDown) {
```

Comment:
I missed that one earlier, you want `if (propagateDown && this.children)`. Otherwise you get the issue you had in unit tests which is an `undefined` exception when trying to iterate over the children.